### PR TITLE
Datapusher fixes

### PR DESCRIPTION
**Pull request by gcsantos-gpa**
Required By

https://github.com/GridProtectionAlliance/SystemCenter/pull/506
✅ https://github.com/GridProtectionAlliance/gpa-gemstone/pull/317

SC-161
SC-162
SC-163
SC-199

### CODE REVIEW

File: Source/Libraries/openXDA.DataPusher/DataPusherRequester.cs
Comment (StephenCWills) [inline Source/Libraries/openXDA.DataPusher/DataPusherRequester.cs:1]: * Copyright 2017?
* Usings are not sorted
* The class is not in a namespace
* The class name has the wrong casing; should be `DataPusherRequester`
* File organization doesn't match convention (use our standard `regions` snippet)
* Private fields `m_Token` and `m_Key` have the wrong casing
* `staticClient` has the wrong naming convention; should be `s_staticClient`
* `async` method names should end with the word `Async`; it should not be placed in the middle of the method name
* `ConfigureRequestAPIRequest()` is a local function; just call it `ConfigureRequest()`
* I count at least three methods that are bypassing this new API and therefore do not support `APIQuery`.
  * `SendMeterConfigurationForInstance()`
  * `SendMeterConfigurationForInstance()`
  * `GetStatus()`

**Review by StephenCWills — COMMENTED** [review]
Do we need to continue to support the legacy `UserAccount` authentication? Seems to me like what we should do is fully switch over to `APIQuery` and then maybe adjust authorization rules to disallow `DataPusher` requests from regular users.

File: Source/Libraries/openXDA.DataPusher/DataPusherRequester.cs
Comment (gcsantos-gpa) [reply Source/Libraries/openXDA.DataPusher/DataPusherRequester.cs:1]: Hey Stephen, thank you for the feedback! I've pushed some commits that should address everything you asked for here.

**Comment by StephenCWills** [comment]
Can we go ahead and create a breaking change then? IMO, we can fix the errors at GTC/GPC easily enough, but we have no process for removing the legacy behavior later.

File: Source/Libraries/openXDA.DataPusher/WebAPIHub.cs [hunk 1]
```diff
@@ -72,588 +46,321 @@ public static string CompanyName
...
-        #region [ Helpers ]
+        #region [ UserAccount Helpers ]
 
-        private static string BuildURL<T>(string baseAddress, string basePath, string whereClause) where T : class
+        private static string BuildPath<T>(string basePath, string whereClause) where T : class
         {
-            string baseURL = BuildURL<T>(baseAddress, basePath);
+            string baseURL = BuildPath<T>(basePath);
             string urlEncodedWhereClause = HttpUtility.UrlEncode(whereClause);
             return $"{baseURL}?where={urlEncodedWhereClause}";
```
Comment (StephenCWills) [inline Source/Libraries/openXDA.DataPusher/WebAPIHub.cs:354]: This pattern has always worried me. We should think about moving away from passing an arbitrary WHERE clause to the server as much as possible so we can retire APIs with high potential for SQL injection.

**Review by StephenCWills — CHANGES_REQUESTED** [review]
Really excellent cleanups in this PR. I like where this is headed. I just noticed a couple problematic patterns that were repeated several times. They should be easy enough to clear up.

File: Source/Libraries/openXDA.DataPusher/WebAPIHub.cs (hunk 1 above)
Comment (StephenCWills) [reply Source/Libraries/openXDA.DataPusher/WebAPIHub.cs:354]: I should also mention, this is just a general comment about `WebAPIHub` to increase awareness. There's nothing actionable here. You can resolve once you've read it.

File: Source/Libraries/openXDA.DataPusher/DataPusherEngine.cs [hunk 2]
```diff
@@ -380,20 +352,144 @@ public void SyncMeterConfigurationForInstance(string clientId, RemoteXDAInstance
...
-                    Asset remoteAsset = AddOrGetRemoteAsset(instance.Address, assetToDataPush, localAsset, localAssetTypes, remoteAssetTypes, userAccount);
-                    if (remoteAsset is null)
-                        return;
-                    connection.ExecuteNonQuery("UPDATE AssetsToDataPush SET Synced = 1 where ID = {0}", assetToDataPush.ID);
+                    // By default, line segments are not associated with meters, so they often get overlooked. Decision was made to always add them if the associated line is here
+                    TableOperations<LineSegment> lineSegmentTable = new TableOperations<LineSegment>(connection);
+                    TableOperations<AssetsToDataPush> assetPushTable = new TableOperations<AssetsToDataPush>(connection);
+                    List<AssetsToDataPush> lineSegments = new TableOperations<Line>(connection).
+                        QueryRecordsWhere("ID in (SELECT LocalXDAAssetID FROM AssetsToDataPush WHERE RemoteXDAInstanceID = {0})", instance.ID).
+                        SelectMany(line =>
```
[the remark this answers is not in the document]
Comment (gcsantos-gpa) [reply Source/Libraries/openXDA.DataPusher/DataPusherEngine.cs:369]: We don't need to but I think this is a good idea, added.

File: Source/Libraries/openXDA.DataPusher/WebAPIHub.cs [hunk 3]
```diff
@@ -72,588 +46,321 @@ public static string CompanyName
...
+            using (HttpResponseMessage response = await requester.SendRequestAsync(path, HttpMethod.Get))
             {
-                ConfigureRequest(request, userAccount);
-
-                using (HttpResponseMessage response = await HttpClient.SendAsync(request))
-                {
-                    if (!response.IsSuccessStatusCode)
-                        throw new InvalidOperationException($"Server returned status code {response.StatusCode}: {response.ReasonPhrase}");
+                if (!response.IsSuccessStatusCode)
+                    throw new InvalidOperationException($"Server returned status code {response.StatusCode}: {response.ReasonPhrase}");
```
[the remark this answers is not in the document]
Comment (gcsantos-gpa) [reply Source/Libraries/openXDA.DataPusher/WebAPIHub.cs:64]: I did so, out of curiosity and ignorance, why is this a bad pattern?

File: Source/Libraries/openXDA.DataPusher/WebAPIHub.cs (hunk 1 above)
Comment (gcsantos-gpa) [reply Source/Libraries/openXDA.DataPusher/WebAPIHub.cs:354]: I too actually have been worried about this exact same thing, so we are of one mind on it. I want to fix it where possible, but it usually isn't actionable as is the case here.

File: Source/Libraries/openXDA.DataPusher/WebAPIHub.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (StephenCWills) [reply Source/Libraries/openXDA.DataPusher/WebAPIHub.cs]: `response.EnsureSuccessStatusCode()` will throw an `HttpResponseException`, I believe. Exception handlers will be more likely to expect it. For example, and if I'm not mistaken, the HTTP status code would be passed through to the client automatically by `ApiController` handlers.

It also reduces nesting, which typically makes the code more readable.

**Review by StephenCWills — APPROVED** [review]

**Pull request merged**